Update for FLSun Hispeedv1 Board

Foxies-CSTL opened the pull request:
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description
Following user feedback, I added a comment section for the UART mode of the TMC220x in serial mode.
An explain to wired the serial mono-wire for the TMC2209.
I fixed the compatibility of Stock TFT displays (MKS Robin TFT).

<!--
We must be able to understand your proposed change from this description. If we can't understand what the code will do from this description, the Pull Request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code recently, so please walk us through the concepts.
-->

### Benefits
 - Compatibility with old Hispeed Board with no removables drivers stepper (A4988).
 - Help for using UART modes for TMC220x.
 - Compatible usage with TMC2209 UART one-wire and module Wifi + other port I/O (M1)
 - Fix SD Card onboard with an extension port and by extention connector J14.
More compatibility with Stock TFT (MKS Robin TFT).
 - Fix the Black screen after flash.
 - Remove XPT2049 part.

<!-- What does this fix or improve? -->

### Configurations
[QQS-Pro.zip](https://github.com/MarlinFirmware/Marlin/files/5465094/QQS-Pro.zip)
[PR 309](https://github.com/MarlinFirmware/Configurations/pull/309)
**Environment VSC/Platformio**
Identical MKS_Robin_nano but the renaming script is that of the MKS_Robin_mini and add "_nanolib_" script to reduce the binary.
**Note:** 
I do not know how to integrate the trick of the script "nanolib" but this one is necessary: Compiling with the Color UI and with UBL leveling results in a binary that is too big.
<!-- Attach any Configuration.h, Configuration_adv.h, or platformio.ini files needed to compile/test your Pull Request. -->

### Related Issues

<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

Foxies-CSTL committed:
Update pins for TMC and UART capacities

Foxies-CSTL committed:
Update QQS_Pro capacities

Foxies-CSTL committed:
Update platformio.ini

Foxies-CSTL committed:
Update pins.h

Foxies-CSTL committed:
Update QQS_Pro Stock/TMC/UART

Foxies-CSTL committed:
Multi_Config FLSUN_QQS-PRO

STOCK, TMC, UART with TFT (320x240) Standard and Marlin UI.

Foxies-CSTL committed:
add for sync

Foxies-CSTL committed:
add for sync

Foxies-CSTL committed:
sync

Foxies-CSTL committed:
sync

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
Update _Statusscreen.h

Foxies-CSTL committed:
sync

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

thinkyhead committed:
Update platformio.ini

thinkyhead committed:
Update platformio.ini

Foxies-CSTL committed:
update QQS-PRO

Foxies-CSTL committed:
Merge branch 'bugfix-2.0.x' of https://github.com/Foxies-CSTL/Marlin_2.0.x into bugfix-2.0.x

Foxies-CSTL committed:
Update platformio.ini

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update Configuration.h

Foxies-CSTL committed:
Update Configuration_adv.h

Foxies-CSTL committed:
Update Configuration.h

Foxies-CSTL committed:
Update Configuration_adv.h

Foxies-CSTL committed:
Update Version.h

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Foxies-CSTL committed:
Update platformio.ini

Foxies-CSTL committed:
Delete _Statusscreen.h

Foxies-CSTL committed:
Delete _Bootscreen.h

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Foxies-CSTL committed:
Update Configuration_adv.h

Foxies-CSTL committed:
Update Configuration.h

Foxies-CSTL committed:
Update add_nanolib.py

thinkyhead committed:
FLSUN Delta QQS-Pro support

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update boards.h

Foxies-CSTL committed:
Merge branch 'bugfix-2.0.x' of https://github.com/Foxies-CSTL/Marlin_2.0.x into bugfix-2.0.x

Foxies-CSTL committed:
Update boards.h

Foxies-CSTL committed:
Update pins.h

Foxies-CSTL committed:
Update pins.h

add hispeed

Foxies-CSTL committed:
Update boards.h

Foxies-CSTL committed:
Update Configuration.h

Remove NUM_SERIAL 2

thinkyhead committed:
Fix bizarre regressions

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Add comment for mode UART serial/hardware.

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Following user feedback, I added a comment section for the UART mode of the TMC220x in serial mode and in hardware mode for the TMC2209.

sjasonsmith commented:
The serial section of this pins file does not seem correct at all. Both the “HARDWARE_SERIAL” and “SOFTWARE_SERIAL” sections are actually using SoftwareSerial.

Is the intent to actually use the RX1/TX1 pins as HardwareSerial (requires two pins), or to use one single pin for all drivers (Using SoftwareSerial)?

Does this board have built-in UART connections, or does the user have to add wires to the stepper drivers? I guess these pins are all on the ESP/WiFi connector?

Foxies-CSTL commented:
Hello,
I will try to be clear
In fact the problem with this board is that it does not have a dedicated connector for the UART. Wiring must be added by the user with only one wire per stepper driver to an RX / TX pin of the WIFI connector.
I just modified the SOFTWARE SERIAL part so that it is the default and that there is no need to activate "#define SOFTWARE_SERIAL" in Configuration.h or in another file anymore. This should work for all TMC220x type stepper drivers in serial mode.
On the other hand, the definition "#define HARDWARE_SERIAL" will have to be activated by the user if he cables all of his drivers step by step on a single pin and this only concerns the TMC2209.
My last modification which should be consistent.
`#if HAS_TMC220x
  
  #if ENABLED(HARDWARE_SERIAL)  /*  TMC2209 */ 
    #define  X_SLAVE_ADDRESS 3    // |  |  :
    #define  Y_SLAVE_ADDRESS 2    // :  |  :
    #define  Z_SLAVE_ADDRESS 1    // |  :  :
    //#define E0_SLAVE_ADDRESS 0    // :  :  :

    #define X_SERIAL_TX_PIN                  PA9  //TXD1
    #define X_SERIAL_RX_PIN                  PA9  //TXD1

    #define Y_SERIAL_TX_PIN                   PA9  //TXD1
    #define Y_SERIAL_RX_PIN                   PA9  //TXD1

    #define Z_SERIAL_TX_PIN                   PA9  //TXD1
    #define Z_SERIAL_RX_PIN                   PA9  //TXD1

/*
 * TMC2208 stepper UART-configurable by PDN_UART pin
 *   Software serial 
 */

  #else  /*  TMC220x   */
    #define  X_SLAVE_ADDRESS 0
    #define  Y_SLAVE_ADDRESS 0
    #define  Z_SLAVE_ADDRESS 0

    #define X_SERIAL_TX_PIN                  PA10  //RXD1
    #define X_SERIAL_RX_PIN                  PA10  //RXD1

    #define Y_SERIAL_TX_PIN                   PA9  //TXD1
    #define Y_SERIAL_RX_PIN                   PA9  //TXD1

    #define Z_SERIAL_TX_PIN                   PC7  //IO1
    #define Z_SERIAL_RX_PIN                   PC7  //IO1

  #endif  
  // Reduce baud rate to improve software serial reliability
  #define TMC_BAUD_RATE 19200
#else`

Thank you for your comments.

sjasonsmith commented:
Thanks for the explanation. The terms SOFTWARE_SERIAL and HARDWARE_SERIAL are used incorrectly here, as normally that refers to whether serial will use the SoftwareSerial or HardwareSerial C++ classes.

This board is doing this different than anything else in Marlin. I don't think we want these `SOFTWARE_SERIAL` and `HARDWARE_SERIAL` defines like this.

Before I can make a recommendation, perhaps @thinkyhead can share his opinion on this. How should TMC serial assignments be defaulted on boards which require users to add custom wiring? I guess we have a few choices:
1. Default to SoftwareSerial on arbitrary pins (the ESP connector in this case)
2. Expect users to modify their pin file (or add to their config file), and leave commented-out suggestions in pins
3. Create board-specific constants as is done here, which require people to discover them through the pins file anyway

From a correctness standpoint, I like number 2. If a user has to add custom wiring, it seems reasonable to expect them to edit a pins file, but maybe you don't agree.

Foxies-CSTL commented:
For the second part (2209 in HARDWARE mode) we could comment it out in the pin definition. So the user will have to modify this file for his specific wiring. This is a good point and I agree.

Foxies-CSTL commented:
I rewrite the definition of the pins and put you here as an attached file.

sjasonsmith commented:
Is this board exactly the same as an MKS Robin Mini?

sjasonsmith commented:
These should be your three different options. I changed a few things from your header.
I don't have actual decision-making authority, so you might want to wait to hear thinkyhead's preference prior to making changes.

1. Removed `SOFTWARE_SERIAL` and `HARDWARE_SERIAL`. These are not used by any other boards. If we want to use a technique like this we should probably plan for it to be usable by all boards.
2. Added a true HardwareSerial option.
3. Removed slave addresses. These are already in Configuration_adv.h, and I think should only be in pins files if hard-wired on a board. We have a sanity check which will produce an error if the user forgets to configure them.

```
#if HAS_TMC220x
  #if HAS_TMC_UART
    // This board has no hard-wired UART pins for TMC drivers.
    // Several wiring options are provided below, defaulting to
    // to the most compatible.

    // Option 1: SoftwareSerial with one pin per driver
    // Compatible with TMC2208 and TMC2209 drivers
    // Insufficient pins for E driver
    #define X_SERIAL_TX_PIN                 PA10  // RXD1
    #define X_SERIAL_RX_PIN                 PA10  // RXD1
    #define Y_SERIAL_TX_PIN                 PA9   // TXD1
    #define Y_SERIAL_RX_PIN                 PA9   // TXD1
    #define Z_SERIAL_TX_PIN                 PC7   // IO1
    #define Z_SERIAL_RX_PIN                 PC7   // IO1
    #define TMC_BAUD_RATE                   19200

    // Option 2: SoftwareSerial with a single pin for four drivers
    // Compatible with TMC2209. Requires SLAVE_ADDRESS definitions
    // in Configuration_adv.h and proper jumper configuration.
    //#define X_SERIAL_TX_PIN                 PA9   // TXD1
    //#define X_SERIAL_RX_PIN                 PA9   // TXD1
    //#define Y_SERIAL_TX_PIN                 PA9   // TXD1
    //#define Y_SERIAL_RX_PIN                 PA9   // TXD1
    //#define Z_SERIAL_TX_PIN                 PA9   // TXD1
    //#define Z_SERIAL_RX_PIN                 PA9   // TXD1
    //#define E0_SERIAL_TX_PIN                PA9   // TXD1
    //#define E0_SERIAL_RX_PIN                PA9   // TXD1
    //#define TMC_BAUD_RATE                   19200

    // Option 3: HardwareSerial with two pins for four drivers
    // Compatible with TMC2209. Provides best performance.
    // Requires SLAVE_ADDRESS definitions in Configuration_adv.h
    // and proper jumper configuration. Uses RXD1 and TXD1 pins
    //#define X_HARDWARE_SERIAL  MSerial1
    //#define Y_HARDWARE_SERIAL  MSerial1
    //#define Z_HARDWARE_SERIAL  MSerial1
    //#define E0_HARDWARE_SERIAL MSerial1    
  #else
```

Foxies-CSTL commented:
I already thought about it during the first definition of the board but with the first part activated (TMC220x), the user has nothing to modify on the definition if he passes his pilots step by step in UART mode in mode serial. I did the same to protect the part with the ESP module. Note: Some people receive the machine with 3xTMC2208 (XYZ) and 1xA4988 (E). That's why I separated the XYZ part from the Extruder part.

sjasonsmith commented:
> Note: Some people receive the machine with 3xTMC2208 (XYZ) and 1xA4988 (E). That's why I separated the XYZ part from the Extruder part.

This distinction belongs in the printer example, not the pins file. Defining the extra E0 elements will do nothing if they don't have a TMC220x enabled in the Configuration.h.

Foxies-CSTL commented:
> 
> 
> Is this board exactly the same as an MKS Robin Mini?

Yes but no ;-) but it is a clone and I had problems with dialogue with the SD card and slowing down of the TFT screen. As soon as I remove the SD card the screen froze !! Some pins like powerloose are set to PA1 while on Robin_mini PA2. This board has fewer free connectors than the Robin_mini.

Foxies-CSTL commented:
> Defining the extra E0 elements will do nothing if they don't have a TMC220x enabled in the Configuration.h.

Yes but only this printer have this board (for the moment) and there is a servo monitoring of the Vref on the stepper pilots. I tried to delete it but the machine goes crazy during some printing. It is a Delta and movement can be destructive.

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Remove 2209 part and  condition for Wifi Module.

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Correction some typo

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_FLSUN_HISPEED.h:
```diff
@@ -109,88 +109,73 @@
+  #define Z_SERIAL_TX_PIN                   PC7   // IO1
+  #define Z_SERIAL_RX_PIN                   PC7   // IO1
+  #define TMC_BAUD_RATE                   19200
+
+  // HardwareSerial with one pins for four drivers
+  // Compatible with TMC2209. Provides best performance.
+  // Requires SLAVE_ADDRESS definitions in Configuration_adv.h
+  // and proper jumper configuration. Uses one I/O pins
+  // like PA10/PA9/PC7/PA8 only.
+  // position the jumpers in this way:  |= close and := open.
+  // ex:
+  //#define  X_SLAVE_ADDRESS 3    // |  |  :
+  //#define  Y_SLAVE_ADDRESS 2    // :  |  :
+  //#define  Z_SLAVE_ADDRESS 1    // |  :  :
+  // For E 0 by default whatever the mode used but remove all jumpers.  
+  
```
Something isn't right in the latest update. You have this comment, but none of the code that goes with it...

Also, I think you have mixed up two of my comments into one.

>   // HardwareSerial with one pins for four drivers
>  // Compatible with TMC2209. Provides best performance.

This is for the version which defines MSerial1 instead of pins. Any mode which defines pins is SoftwareSerial, and is not "best performance".

Foxies-CSTL replied:
> Also, I think you have mixed up two of my comments into one.

Agree on the terms but when we validate an assembly of four stepper drivers (TMC2209) in UART linked to the same serial pin and we define their communication address by the position of the jumpers. Isn't that a hardware mode?

sjasonsmith replied:
SoftwareSerial = Uses any I/O pin, and serial bit timing is manually controlled by software. Timing is very inaccurate, so low baud rates help make it more reliable.

HardwareSerial = Uses UART peripheral in the MCU. Accurate bit timing, but only works on the correct RX/TX pins.
                             User does not specify pin numbers, just a serial port name (like `MSerial1`)

Slave addresses can be used with TMC2209 drivers with both SoftwareSerial or HardwareSerial.

Foxies-CSTL replied:
Aaah, Ok!! Thank you, it's clearer now, sorry for the inconvenience.

thinkyhead committed:
Update pins_FLSUN_HISPEED.h

thinkyhead committed:
Update pins_FLSUN_HISPEED.h

thinkyhead committed:
More compatible driver macros

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

#19956

Foxies-CSTL committed:
Fix SPI2-Extension SDCard-TFT buz

Fix I/O with extension SDCard
Fix  Buzzer aren't Speaker into the TFT screen.

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Change Wifi pin for compatibility for other modules ESP-WFI.
Fix XPT2046_CALIBRATION.
Some cleanings and infos.

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_FLSUN_HISPEED.h:
```diff
@@ -106,49 +106,39 @@
-
-    #define Z_SERIAL_TX_PIN                 PC7   // IO1
-    #define Z_SERIAL_RX_PIN                 PC7   // IO1
+  /**
```
I think you should simply delete this section.

Slave addresses already exist in Configuration_adv.h, and are not included in other pins files.

Is there a common scenario where they are used on this board, or is this just for the occasional user with their own custom configuration?

We already have checks at build time to be sure they cannot forget to set them if multiple drivers use the same pins.

Foxies-CSTL replied:
Hi, I have updated some fixes and simplification in some parts:
TMC remove serial address
Added the J2 connector and the connector for the SD extension card.
Collapsed XPT2046 into one block for all three UI types.
Some typographies.

Thank you and sorry for the wait, I validated the correct operation of the TMC2209 stepper driver in single-wire UART mode.

sjasonsmith reviewed (commented):
This is my only outstanding concern, aside from minor cosmetic/formatting issues.

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Simplication and add J2 connector

Reformatage differents parts:
-TMC2209 Serial one wire
-XPT20046 Touch Screen
-SdCard and card socket

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

Foxies-CSTL committed:
Update pins_FLSUN_HISPEED.h

Better integration managment SDCard and extension

Foxies-CSTL committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

thinkyhead committed:
Update pins_FLSUN_HISPEED.h

thinkyhead committed:
Update pins_FLSUN_HISPEED.h

Foxies-CSTL committed:
Update following the new management of screens.

Removal of part XPT2046.
Displacement of the TFT buffer in the FSMC part and of the colors specific to the LCD emulation (UI Classic).
